### PR TITLE
JSON-LD Context generator

### PR DESCRIPTION
**Pull request by DiegoPino**
Resolves https://github.com/Islandora-CLAW/CLAW/issues/493 
Resolves https://github.com/Islandora-CLAW/CLAW/issues/505 (via caching and cache tags)
update:
**UPDATE:**

Ready to be reviewed and merged

**UPDATE:**

Resolves https://github.com/Islandora-CLAW/CLAW/issues/506 (via controller that responds with application/ld+json)

Base pull here, service, interface, and Class. Missing tests. Some edge cases not covered, probably more complicated than it should. Drupal Coding standards cleaned.
Issue 505 solved via automatic cache invalidation via cache tags.

~~Please, DO NOT MERGE!. I repeat, please DO NOT MERGE!~~ 😝 

~~Update: Ok, now this should do something and could be eventually merged~~

Testing:

if you enable this module and go to
http://localhost:8000/fedora_resource_context/rdf_source
or
http://localhost:8000/fedora_resource_context/rdf_non_source
or
http://localhost:8000/fedora_resource_context/collection
You should see some `@context`being generated

If you feel lucky, you can add more mappings to our YAML bundle mappings and see those changes reflected (remember that config entities are cached, so run always `drupal cache:rebuild`)

Please inform of any issues so I can keep improving this service.

Thanks!

### CODE REVIEW

**Comment by whikloj** [comment]
@dannylamb Looking at how @DiegoPino has implemented this, it would appear the simple implementation of a BundleResolver is just a `str_replace()` to turn the context url to the bundle. 

I'll let you look at this and if that is what you need, I still up to writing the Interface and simple bundle resolver.

File: src/RdfBundleSolver/JsonldContextGenerator.php [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+      $jsonLdContextArray['@context'] = $theAccumulator;
+    }
+    return json_encode($jsonLdContextArray, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
+
+  }
+
+  /**
+   * Get RDF types for all bundles associated to a content type.
+   *
+   * This function is not used right now, but will be needed to be able to
```
Comment (dannylamb) [inline src/RdfBundleSolver/JsonldContextGenerator.php]: If this isn't used yet, then it's not neccessary for this PR.

File: src/RdfBundleSolver/JsonldContextGenerator.php [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+   *    Rdf mapping object.
+   * @param string $cid
+   *   Name of the cache bin to use.
+   *
+   * @return string
+   *   A json encoded string for the processed JSON-LD @context
+   */
+  protected function writeCache(RdfMappingInterface $rdfMapping, $cid) {
+
+    $data = $this->generateContext($rdfMapping);
```
Comment (dannylamb) [inline src/RdfBundleSolver/JsonldContextGenerator.php]: We either need to try/catch here or formalize that we're passing the exception through with an `@throws` annotation.

File: src/Controller/FedoraResourceJsonLdContextController.php [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ * @package Drupal\islandora\Controller
+ */
+class FedoraResourceJsonLdContextController extends ControllerBase {
+
+  /**
+   * Displays JSON-LD Context for a fedora_resource bundle.
+   */
+  public function content($bundle, Request $request) {
+
+    $context = \Drupal::service('islandora.jsonldcontextgenerator')->getContext('fedora_resource.' . $bundle);
```
Comment (DiegoPino) [inline src/Controller/FedoraResourceJsonLdContextController.php]: Good, that is how I would make it if this piece of code would stay (it wont) Jinx, we were working that link yesterday with @whikloj . The whole Context Controller will be removed before the merge. Only there to allow people to test the context.

File: src/Controller/FedoraResourceJsonLdContextController.php [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    $build = array(
+      '#type' => 'markup',
+      '#prefix' => '<pre>',
+      '#suffix' => '</pre>',
+      '#markup' => $context,
+    );
+    return $build;
```
Comment (DiegoPino) [inline src/Controller/FedoraResourceJsonLdContextController.php]: same as above. Not part of the pull nor the tickets, just code to enable people to test.
Seems like you want the controller to stay? If so, the Path needs to be changed and of course the injection via __create and construct(). So should i make it stay?

File: src/RdfBundleSolver/JsonldContextGenerator.php [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+   * @return string
+   *    The JSON-LD @context in string form.
+   */
+  public function getContext($ids = 'fedora_resource.rdf_source') {
+    $cid = JsonldContextGenerator::CACHE_BASE_CID . $ids;
+    $cache = $this->cache->get($cid);
+    $data = '';
+    if (!$cache) {
+      $rdfMapping = RdfMapping::load($ids);
+      if ($rdfMapping) {
```
Comment (DiegoPino) [inline src/RdfBundleSolver/JsonldContextGenerator.php]: It is handling that:  `RdfMapping::load` returns `NULL` if failed,
```PHP
/**
   * Loads one entity.
   *
   * @param mixed $id
   *   The ID of the entity to load.
   *
   * @return \Drupal\Core\Entity\EntityInterface|null
   *   An entity object. NULL if no matching entity is found.
   */
  public function load($id);
```
 the `if ` checks for that.

File: src/RdfBundleSolver/JsonldContextGenerator.php (hunk 1 above)
Comment (DiegoPino) [inline src/RdfBundleSolver/JsonldContextGenerator.php]: @dannylamb this is a collaborative work with @whikloj. I made that method to facilitate the work of a bundle resolver and because this is the class and the place where such a method should belong. Also. this is the first approach for the context service, that function will be helpfull to make the @context generation more robust (like being able to define the @type for a referenced URI) and I don't want to write it again. I don't see the harm but if there is consensus i can remove it. @whikloj maybe you have not use for this either?

File: src/RdfBundleSolver/JsonldContextGenerator.php (hunk 2 above)
Comment (DiegoPino) [inline src/RdfBundleSolver/JsonldContextGenerator.php]: true. Will do today

File: src/RdfBundleSolver/JsonldContextGeneratorInterface.php [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+ * @ingroup: islandora
+ */
+interface JsonldContextGeneratorInterface {
+
+  /**
+   * Generates an JSON-LD Context string based on an RdfMapping object.
+   *
+   * @param \Drupal\rdf\Entity\RdfMapping|RdfMappingInterface $mapping
+   *    An RDF Mapping Object.
+   *
```
Comment (DiegoPino) [inline src/RdfBundleSolver/JsonldContextGeneratorInterface.php]: ok

File: src/RdfBundleSolver/JsonldContextGenerator.php (hunk 1 above)
Comment (dannylamb) [inline src/RdfBundleSolver/JsonldContextGenerator.php]: I'd prefer this be included in a PR for the functionality it addresses, so that it appears in the diff and gets reviewed at the appropriate time.

File: src/Controller/FedoraResourceJsonLdContextController.php (hunk 4 above)
Comment (dannylamb) [inline src/Controller/FedoraResourceJsonLdContextController.php]: Figured you were going for https://github.com/Islandora-CLAW/CLAW/issues/506 while you were at it.  That will bring on its own baggage, though.  The controller will need tests too.

Personally, I'd just cut the scope of this and focus on what's in Islandora-CLAW/CLAW#493 and Islandora-CLAW/CLAW#505.  We can always use the devel block or `drush scr` to test.

**Comment by DiegoPino** [comment]
@whikloj @dannylamb lots of changes and concerns addressed (I hope).

- Controller responds with application/ld+json, so no HTML, look at your downloads when testing the routes
- Naive Drupal to JSON-LD term definition mapping used to set default datatypes
- throws exceptions
- try/catches the previous ones
- docs update
- new service to manage logging for islandora modules (islandora channel)
- Removed the bundle/entity resolver helper method as requested

NO TEST until monday. So much new code. Could not write test.

File: src/RdfBundleSolver/JsonldContextGenerator.php [hunk 7]
```diff
@@ -250,8 +245,19 @@ private function getFieldsRdf(RdfMappingInterface $rdfMapping, $field_name, Fiel
    */
   protected function writeCache(RdfMappingInterface $rdfMapping, $cid) {
 
-    $data = $this->generateContext($rdfMapping);
-    $this->cache->set($cid, $data, Cache::PERMANENT, $rdfMapping->getCacheTagsToInvalidate());
+    // This is how an empty json encoded @context looks like.
+    $data = json_encode(array('@context' => ''), JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
```
Comment (whikloj) [inline src/RdfBundleSolver/JsonldContextGenerator.php]: Should we use `JSON_PRETTY_PRINT` here or should we keep the json small and tightly packed? 

I wondered if `cache->set()` would pack it up, but in case it doesn't perhaps we should keep it smaller to save (a little) space?

Comment (DiegoPino) [inline src/RdfBundleSolver/JsonldContextGenerator.php]: @whikloj true, thanks for reviewing 😄. 
True.
Suggestion: Let's give it a few rounds to make sure the JSON-LD @contexts are well formed and good to go, which will happen when we start using this. The pretty print makes it simpler to debug. Once we start using this and we are comfortable I promise I remove the pretty print. 👍 (and open an issue for that).
Most of my coding and comments are somehow inclined to help others jumping into debugging or even implementing the code easier. I know it's sometimes annoying, and also superfluos, but with so little people being able to iterate on everything, any little error on logic can become complicated.

**Comment by dannylamb** [comment]
:+1: These changes look great @DiegoPino.  Feel free to tag me if you need help setting up those tests.

**Comment by DiegoPino** [comment]
@dannylamb thanks! 😀  No worries, have the tests under control.  It is just that testing controllers is a pain - long - not complicated and I'm on a train to Boston. NO worries, enjoy your vacations. Will have those for Monday.

**Comment by DiegoPino** [comment]
@ruebot both: coming back (and rebasing) and occupied with release duties. Thanks

**Comment by DiegoPino** [comment]
@ruebot will have when committing. Do you need that commit now? This is going to be my last pull for some time, so want to have it right.

**Comment by DiegoPino** [comment]
Some updates: I'm having some issues with tests and strictness of `phpcs` lately, but should be able to commit a working and clean version by today. I see other pull requests that need to be merged, so will probably need a new rebase here also.

**Comment by DiegoPino** [comment]
Some changes... (will have to rebase but this is working)

Cache
Now caching happens on the response and on the method. Best of both
worlds.
```Shell
curl -i http://localhost:8000/fedora_resource_context/rdf_source
HTTP/1.1 200 OK
Date: Tue, 21 Mar 2017 19:19:03 GMT
Server: Apache/2.4.18 (Ubuntu)
Cache-Control: must-revalidate, no-cache, private
X-Powered-By: Islandora CLAW API
X-Drupal-Dynamic-Cache: MISS
X-UA-Compatible: IE=edge
Content-language: en
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Expires: Sun, 19 Nov 1978 05:00:00 GMT
X-Generator: Drupal 8 (https://www.drupal.org)
X-Debug-Token: 7d33c2
X-Debug-Token-Link: /admin/reports/profiler/view/7d33c2
X-Drupal-Cache: HIT
Content-Length: 229
Content-Type: application/ld+json

{"@context":{"schema":"http://schema.org/","schema:dateModified":{"@type
":"xsd:dateTime"},"schema:dateCreated":{"@type":"xsd:dateTime"},"fedora"
:"http://fedora.info/definitions/v4/repository#","fedora:hasParent":{"@t
ype":"@id"}}}
````

And after cache rebuild, user permissions changed, bundle updated, etc:

```Shell
HTTP/1.1 200 OK
Date: Tue, 21 Mar 2017 19:20:49 GMT
Server: Apache/2.4.18 (Ubuntu)
Cache-Control: must-revalidate, no-cache, private
X-Powered-By: Islandora CLAW API
X-Drupal-Dynamic-Cache: MISS
X-UA-Compatible: IE=edge
Content-language: en
X-Content-Type-Options: nosniff
X-Frame-Options: SAMEORIGIN
Expires: Sun, 19 Nov 1978 05:00:00 GMT
X-Generator: Drupal 8 (https://www.drupal.org)
X-Debug-Token: acc399
X-Debug-Token-Link: /admin/reports/profiler/view/acc399
X-Drupal-Cache: MISS
Content-Length: 229
Content-Type: application/ld+json
````

**Comment by DiegoPino** [comment]
To run Web Tests manually from inside vagrant machine:
```Shell
sudo -u www-data php core/scripts/run-tests.sh --verbose --class
"Drupal\islandora\Tests\Web\JsonldContextGeneratorWebTest"
```
They are extremely simple. Probably should do all type of crazy tests.
Will wait for this to pass Travis-CI and then will rebase and then will commit the Kernel ones

**Comment by DiegoPino** [comment]
@dannylamb @ruebot @whikloj there: kernel tests + web tests + travis fixed to run correct drush server and allow testing. Please review. thanks

File: .travis.yml [hunk 8]
```diff
@@ -22,18 +21,10 @@ install:
...
+  - cd web; drush --uri=127.0.0.1:8282 en -y islandora
 
 script:
   - $SCRIPT_DIR/line_endings.sh $TRAVIS_BUILD_DIR
   - phpcs --standard=Drupal --extensions=php,module,inc,install,test,profile,theme,css,info $TRAVIS_BUILD_DIR
   - phpcpd --names *.module,*.inc,*.test,*.php $TRAVIS_BUILD_DIR
-  - php core/scripts/run-tests.sh --verbose --php `which php` islandora
-
-notifications:
-  irc:
```
Comment (ruebot) [inline .travis.yml]: This shouldn't be removed.

Comment (DiegoPino) [inline .travis.yml]: I know, i don't keep that in my local travis-CI integration and forgot to add back. restored

Comment (DiegoPino) [inline .travis.yml]: @ruebot any other changes? not sure if you go the notification so checking

File: .travis.yml [hunk 9]
```diff
@@ -1,7 +1,6 @@
 sudo: true
 language: php
 php:
-  - 5.5
```
Comment (ruebot) [inline .travis.yml]: Why are you committing this? This comment needs to be removed, and replaced with https://github.com/Islandora-CLAW/islandora/commit/779a9ee01859dee6784fefb018ccabf2fe7e0421

Comment (DiegoPino) [inline .travis.yml]: I made that change _while_ you were discussing removing PHP5, just to avoid to have to REBASE again. What about marking all my errors in one pass instead every 24 hours in the same file? Hope not asking too much. thanks!

**Comment by DiegoPino** [comment]
@ruebot @whikloj @dannylamb: Rebased again and force pushed my local branch to preserve that last @ruebot commit/comment on the PHP5.5 removal thing.

File: src/RdfBundleSolver/JsonldContextGenerator.php [hunk 10]
```diff
@@ -0,0 +1,424 @@
...
+      $data = $cache->data;
+    }
+    return $data;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function generateContext(RdfMappingInterface $rdfMapping) {
+    // TODO: we will need to use \Drupal\Core\Fieldt\FieldDefinitionInterface
```
Comment (dannylamb) [inline src/RdfBundleSolver/JsonldContextGenerator.php]: s/Fieldt/Field

File: src/RdfBundleSolver/JsonldContextGeneratorInterface.php [hunk 11]
```diff
@@ -0,0 +1,44 @@
+<?php
+
+namespace Drupal\islandora\RdfBundleSolver;
```
Comment (DiegoPino) [inline src/RdfBundleSolver/JsonldContextGeneratorInterface.php]: @dannylamb there is a feasible/reasonable reason for doing this, naming this like that. Not now, but after MVP we could need a real rdf:type to bundle solving service. building our API based on the idea that the client needs to know the internals of our Drupal installation and match the bundle names is good for now but not scalable. The context generator should be really just the first step in achieving this. Who knows, Framing and RDF type solving and even SPARQL-UPDATE functionality for PATCH could go into this namespace. But, if you need me to change it, of course i can change it.

File: src/RdfBundleSolver/JsonldContextGenerator.php (hunk 10 above)
Comment (DiegoPino) [inline src/RdfBundleSolver/JsonldContextGenerator.php]: Ok, will fix the todo typo

**Comment by DiegoPino** [comment]
Also, to test suit (Module "Testing") needs to be enabled (if you have a fresh vagrant machine)

**Comment by dannylamb** [comment]
YES, it sure does.  That sounds like a good newbie ticket.

**Comment by dannylamb** [comment]
_ceremoniously waits for Travis_

**Comment by dannylamb** [comment]
Thank you for your contribution @DiegoPino
